library/psa_crypto.c

```diff
@@ -3887,7 +3874,707 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
+    }
+
+    if( operation->lengths_set )
+    {
+        if( operation->ad_remaining != 0 || operation->body_remaining != 0 )
+        {
+            return( PSA_ERROR_BAD_STATE );
+        }
+    }
+
+    if( tag_length < PSA_AEAD_TAG_LENGTH( operation->alg ) )
+    {
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+    }
+
+    finish_output_size = PSA_AEAD_FINISH_OUTPUT_SIZE(operation->alg);
+
+    if( plaintext_size < finish_output_size )
+    {
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+    }
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)
+    if( operation->alg == PSA_ALG_GCM )
+    {
+        if( operation->buffer_length )
+        {
+
+            status =  mbedtls_to_psa_error( mbedtls_gcm_update( &operation->ctx.gcm,
+                                                                PSA_BLOCK_CIPHER_BLOCK_MAX_SIZE,
+                                                                operation->body_buffer,
+                                                                plaintext ) );
+
+            if( status != PSA_SUCCESS )
+            {
+                return( status );
+            }
+
+            *plaintext_length = finish_output_size;
+        }
+
+        // Call finish to get the tag for comparison
+        status =  mbedtls_to_psa_error( mbedtls_gcm_finish( &operation->ctx.gcm,
+                                                            check_tag,
+                                                            16 ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)
+    if( operation->alg == PSA_ALG_CCM )
+    {
+        /* Multipart CCM not supported as yet */
+        ( void ) plaintext;
+        ( void ) plaintext_size;
+        ( void ) plaintext_length;
+        ( void ) tag;
+        ( void ) tag_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+    if( operation->alg == PSA_ALG_CHACHA20_POLY1305 )
+    {
+        status = mbedtls_to_psa_error( mbedtls_chachapoly_finish( &operation->ctx.chachapoly,
+                                                                  check_tag ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305 */
+    {
+        ( void ) plaintext;
+        ( void ) plaintext_size;
+        ( void ) plaintext_length;
+        ( void ) tag;
+        ( void ) tag_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    if( status == PSA_SUCCESS )
+    {
+         /* Check tag in "constant-time" */
+        for( i = 0; i < tag_length; i++ )
+        {
+            tag_diff |= tag[i] ^ check_tag[i];
+        }
+
+        if( tag_diff != 0 )
+        {
+            status = MBEDTLS_ERR_GCM_AUTH_FAILED;
+        }
+    }
+
+    psa_aead_abort_internal(operation);
+
+    return ( status );
+}
+
+/* Abort a multipart AEAD operation */
+psa_status_t psa_aead_abort(psa_aead_operation_t *operation)
```

Comment:
Might as well rename `psa_aead_abort_internal` and change it to return `PSA_SUCCESS`.